pkg/relayer/session/proof.go

```diff
@@ -94,7 +98,11 @@ func (rs *relayerSessionsManager) waitForEarliestSubmitProofsHeightAndGeneratePr
 	// submitProofWindowStartBlock is the block that will have its hash used as the
 	// source of entropy for all the session trees in that batch, waiting for it to
 	// be received before proceeding.
-	submitProofWindowStartBlock := rs.waitForBlock(ctx, submitProofsWindowStartHeight)
+	sessionPathBlock := rs.waitForBlock(ctx, sessionGracePeriodEndHeight)
```

Comment:
Not sure this wait is effective, `supplierClient.CreateClaim` already waits for the tx to be committed.

I also think we need a better definition of what should be the `sessionPathBlock` and from which block it needs to be derived from. I believe it should be in terms of `submitProofWindowStartHeight`.

Reply:
Nice observation! I see your point. :thinking: 

It's especially interesting because this change is what fixed the session E2E test that was failing due to invalid proof paths. I'm going to do some more investigating and come back with more info.

Reply:
For additional context: The issue this is trying to resolve is that we only get a reference to the block we're waiting for (since we threw out the block query client). Since the block whose hash is used for the proof path seed comes earlier than the block we're actually waiting for here (proof window begin), we need another way get that earlier block hash.

Reply:
@red-0ne and I had a discussion about this, we think we understand why this is working and have a plan to make it more explicit. 

### Background

The reason we're seeing the "old blocks" in the range is because of an improvement that we made a while back, which populates the `ReplayClient#EventsSequence()` return replay observable's replay buffer, starting from the time of  `ReplayClient` construction (as opposed to starting from when `#EventSequence()` is called). These "old blocks" are observed by the range because they're in the replay buffer, which is emitted by new replay observable subscribers, prior to emitting real-time notifications (not from the replay buffered).

The size of the replay observable's replay buffer sets a limit on how many past blocks (in this case) are observed by `#EventsSequence()` callers (and derivatives thereof; e.g. `#CommittedBlocksSequence`), before being "caught up" and notified of in real-time.

### Proposal
_(In another PR)_

Make `#CommittedBlocksSequence()` and `#EventsSequence()` receive an argument (`replayBufferStartOffset`) which limits the earliest replay buffer notification emitted. This also applies the the analogous `DelegationClient` method.

In this case, the new `replayBufferStartOffset` MUST be large enough such that it ensures that the path block height for the earliest session in the session trees map. In order to satisfy this constraint, in the event that the earliest session path height block was not observed (e.g. cold start w/ old persisted session trees), we also need to add support in the block client to pre-populate the replay buffer by querying with a `BlockQueryClient`, starting from the earliest late session end height.

---

Another bug we exposed is how `mapBlockToSessionsToClaim()` is currently emitting on-time sessions, in the event that late sessions are present. As currently written, on-time sessions would be emitted to observers one block after late sessions. This map is intended to categorize late vs. on-time sessions by grouping them into separate downstream notifications; however, altering the timing of those notifications was not intended. 

To correct for this we need to (re-)re-write the `mapBlockToSessionsToClaim()` `Map()` as a `ForEach()`. @red-0ne had expressed it this way in [a previous iteration](https://github.com/pokt-network/poktroll/blob/7c69ed8aa792d685bfaa0758050c7f625434001a/pkg/relayer/session/session.go#L90).

Reply:
All settled!